File: llama_mobile-capacitor-plugin/android/src/main/java/com/capacitor/plugin/llamamobile/LlamaMobilePlugin.java
package com.capacitor.plugin.llamamobile;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONException;

@CapacitorPlugin(name = "LlamaMobile")
public class LlamaMobilePlugin extends Plugin {

    private LlamaMobileImpl implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new LlamaMobileImpl(getContext());
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        JSObject params = call.getObject("params");
        if (params == null) {
            call.reject("Missing params");
            return;
        }

        try {
            boolean success = implementation.initialize(params);
            JSObject result = new JSObject();
            result.put("success", success);
            call.resolve(result);
        } catch (JSONException e) {
            call.reject("JSON error: " + e.getMessage());
        }
    }

    @PluginMethod
    public void generate(PluginCall call) {
        JSObject params = call.getObject("params");
        if (params == null) {
            call.reject("Missing params");
            return;
        }

        try {
            JSObject result = implementation.generate(params);
            if (result != null) {
                call.resolve(result);
            } else {
                call.reject("Generation failed");
            }
        } catch (JSONException e) {
            call.reject("JSON error: " + e.getMessage());
        }
    }

    @PluginMethod
    public void multimodalCompletion(PluginCall call) {
        JSObject params = call.getObject("params");
        JSArray mediaPathsArray = call.getArray("mediaPaths");

        if (params == null || mediaPathsArray == null) {
            call.reject("Missing params or mediaPaths");
            return;
        }

        List<String> mediaPaths = new ArrayList<>();
        try {
            for (int i = 0; i < mediaPathsArray.length(); i++) {
                mediaPaths.add(mediaPathsArray.getString(i));
            }

            JSObject result = implementation.multimodalCompletion(params, mediaPaths);
            if (result != null) {
                call.resolve(result);
            } else {
                call.reject("Generation failed");
            }
        } catch (JSONException e) {
            call.reject("JSON error: " + e.getMessage());
        }
    }

    @PluginMethod
    public void stopCompletion(PluginCall call) {
        implementation.stopCompletion();
        call.resolve();
    }

    @PluginMethod
    public void tokenize(PluginCall call) {
        String text = call.getString("text");
        if (text == null) {
            call.reject("Missing text");
            return;
        }

        List<Integer> tokens = implementation.tokenize(text);
        if (tokens != null) {
            JSObject result = new JSObject();
            result.put("tokens", tokens);
            call.resolve(result);
        } else {
            call.reject("Tokenization failed");
        }
    }

    @PluginMethod
    public void detokenize(PluginCall call) {
        JSArray tokensArray = call.getArray("tokens");
        if (tokensArray == null) {
            call.reject("Missing tokens");
            return;
        }

        List<Integer> tokens = new ArrayList<>();
        try {
            for (int i = 0; i < tokensArray.length(); i++) {
                tokens.add(tokensArray.getInt(i));
            }

            String text = implementation.detokenize(tokens);
            if (text != null) {
                JSObject result = new JSObject();
                result.put("text", text);
                call.resolve(result);
            } else {
                call.reject("Detokenization failed");
            }
        } catch (JSONException e) {
            call.reject("JSON error: " + e.getMessage());
        }
    }

    @PluginMethod
    public void generateEmbeddings(PluginCall call) {
        String text = call.getString("text");
        if (text == null) {
            call.reject("Missing text");
            return;
        }

        List<Double> embeddings = implementation.generateEmbeddings(text);
        if (embeddings != null) {
            JSObject result = new JSObject();
            result.put("embeddings", embeddings);
            call.resolve(result);
        } else {
            call.reject("Embedding generation failed");
        }
    }

    @PluginMethod
    public void applyLoraAdapters(PluginCall call) {
        JSArray adaptersArray = call.getArray("adapters");
        if (adaptersArray == null) {
            call.reject("Missing adapters");
            return;
        }

        List<JSObject> adapters = new ArrayList<>();
        try {
            for (int i = 0; i < adaptersArray.length(); i++) {
                adapters.add(adaptersArray.getJSObject(i));
            }

            boolean success = implementation.applyLoraAdapters(adapters);
            JSObject result = new JSObject();
            result.put("success", success);
            call.resolve(result);
        } catch (JSONException e) {
            call.reject("JSON error: " + e.getMessage());
        }
    }

    @PluginMethod
    public void removeLoraAdapters(PluginCall call) {
        implementation.removeLoraAdapters();
        call.resolve();
    }

    @PluginMethod
    public void initMultimodal(PluginCall call) {
        String mmprojPath = call.getString("mmprojPath");
        Boolean useGpu = call.getBoolean("useGpu");

        if (mmprojPath == null || useGpu == null) {
            call.reject("Missing mmprojPath or useGpu");
            return;
        }

        boolean success = implementation.initMultimodal(mmprojPath, useGpu);
        JSObject result = new JSObject();
        result.put("success", success);
        call.resolve(result);
    }

    @PluginMethod
    public void isMultimodalEnabled(PluginCall call) {
        boolean isEnabled = implementation.isMultimodalEnabled();
        JSObject result = new JSObject();
        result.put("enabled", isEnabled);
        call.resolve(result);
    }

    @PluginMethod
    public void releaseMultimodal(PluginCall call) {
        implementation.releaseMultimodal();
        call.resolve();
    }

    @PluginMethod
    public void generateResponse(PluginCall call) {
        String userMessage = call.getString("userMessage");
        Integer maxTokens = call.getInt("maxTokens");

        if (userMessage == null || maxTokens == null) {
            call.reject("Missing userMessage or maxTokens");
            return;
        }

        JSObject result = implementation.generateResponse(userMessage, maxTokens);
        if (result != null) {
            call.resolve(result);
        } else {
            call.reject("Response generation failed");
        }
    }

    @PluginMethod
    public void clearConversation(PluginCall call) {
        implementation.clearConversation();
        call.resolve();
    }

    @PluginMethod
    public void getGrammarContent(PluginCall call) {
        String grammarName = call.getString("grammarName");
        if (grammarName == null) {
            call.reject("Missing grammarName");
            return;
        }

        String content = implementation.getGrammarContent(grammarName);
        if (content != null) {
            JSObject result = new JSObject();
            result.put("content", content);
            call.resolve(result);
        } else {
            call.reject("Grammar not found");
        }
    }

    @PluginMethod
    public void release(PluginCall call) {
        implementation.release();
        call.resolve();
    }
}

File: llama_mobile-capacitor-plugin/ios/Sources/LlamaMobilePlugin/LlamaMobilePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LlamaMobilePlugin)
public class LlamaMobilePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LlamaMobilePlugin"
    public let jsName = "LlamaMobile"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "multimodalCompletion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopCompletion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "tokenize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detokenize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generateEmbeddings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "applyLoraAdapters", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeLoraAdapters", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initMultimodal", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isMultimodalEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseMultimodal", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generateResponse", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearConversation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getGrammarContent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "release", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = LlamaMobile()

    @objc func initialize(_ call: CAPPluginCall) {
        guard let params = call.getObject("params") else {
            call.reject("Missing params")
            return
        }
        
        let result = implementation.initialize(params)
        call.resolve([
            "success": result
        ])
    }
    
    @objc func generate(_ call: CAPPluginCall) {
        guard let params = call.getObject("params") else {
            call.reject("Missing params")
            return
        }
        
        if let result = implementation.generate(params) {
            call.resolve(result)
        } else {
            call.reject("Generation failed")
        }
    }
    
    @objc func multimodalCompletion(_ call: CAPPluginCall) {
        guard let params = call.getObject("params"),
              let mediaPaths = call.getArray("mediaPaths") as? [String] else {
            call.reject("Missing params or mediaPaths")
            return
        }
        
        if let result = implementation.multimodalCompletion(params, mediaPaths: mediaPaths) {
            call.resolve(result)
        } else {
            call.reject("Generation failed")
        }
    }
    
    @objc func stopCompletion(_ call: CAPPluginCall) {
        implementation.stopCompletion()
        call.resolve()
    }
    
    @objc func tokenize(_ call: CAPPluginCall) {
        guard let text = call.getString("text") else {
            call.reject("Missing text")
            return
        }
        
        if let tokens = implementation.tokenize(text) {
            call.resolve([
                "tokens": tokens
            ])
        } else {
            call.reject("Tokenization failed")
        }
    }
    
    @objc func detokenize(_ call: CAPPluginCall) {
        guard let tokens = call.getArray("tokens") as? [Int] else {
            call.reject("Missing tokens")
            return
        }
        
        if let text = implementation.detokenize(tokens) {
            call.resolve([
                "text": text
            ])
        } else {
            call.reject("Detokenization failed")
        }
    }
    
    @objc func generateEmbeddings(_ call: CAPPluginCall) {
        guard let text = call.getString("text") else {
            call.reject("Missing text")
            return
        }
        
        if let embeddings = implementation.generateEmbeddings(text) {
            call.resolve([
                "embeddings": embeddings
            ])
        } else {
            call.reject("Embedding generation failed")
        }
    }
    
    @objc func applyLoraAdapters(_ call: CAPPluginCall) {
        guard let adapters = call.getArray("adapters") as? [[String: Any]] else {
            call.reject("Missing adapters")
            return
        }
        
        let success = implementation.applyLoraAdapters(adapters)
        call.resolve([
            "success": success
        ])
    }
    
    @objc func removeLoraAdapters(_ call: CAPPluginCall) {
        implementation.removeLoraAdapters()
        call.resolve()
    }
    
    @objc func initMultimodal(_ call: CAPPluginCall) {
        guard let mmprojPath = call.getString("mmprojPath"),
              let useGpu = call.getBool("useGpu") else {
            call.reject("Missing mmprojPath or useGpu")
            return
        }
        
        let success = implementation.initMultimodal(mmprojPath, useGpu: useGpu)
        call.resolve([
            "success": success
        ])
    }
    
    @objc func isMultimodalEnabled(_ call: CAPPluginCall) {
        let isEnabled = implementation.isMultimodalEnabled()
        call.resolve([
            "enabled": isEnabled
        ])
    }
    
    @objc func releaseMultimodal(_ call: CAPPluginCall) {
        implementation.releaseMultimodal()
        call.resolve()
    }
    
    @objc func generateResponse(_ call: CAPPluginCall) {
        guard let userMessage = call.getString("userMessage"),
              let maxTokens = call.getInt("maxTokens") else {
            call.reject("Missing userMessage or maxTokens")
            return
        }
        
        if let result = implementation.generateResponse(userMessage, maxTokens: maxTokens) {
            call.resolve(result)
        } else {
            call.reject("Response generation failed")
        }
    }
    
    @objc func clearConversation(_ call: CAPPluginCall) {
        implementation.clearConversation()
        call.resolve()
    }
    
    @objc func getGrammarContent(_ call: CAPPluginCall) {
        guard let grammarName = call.getString("grammarName") else {
            call.reject("Missing grammarName")
            return
        }
        
        if let content = implementation.getGrammarContent(grammarName) {
            call.resolve([
                "content": content
            ])
        } else {
            call.reject("Grammar not found")
        }
    }
    
    @objc func release(_ call: CAPPluginCall) {
        implementation.release()
        call.resolve()
    }
}
